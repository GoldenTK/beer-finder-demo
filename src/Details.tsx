import React, { FC, useEffect, useState } from 'react'
import { Beer } from './types'
import callApi, { RequestMethod } from './api'
import Toast from './Toast'
import { CircularProgress, Typography, Card, CardContent, CardMedia, makeStyles, createStyles, AppBar, Toolbar, IconButton, Paper, } from '@material-ui/core'
import { useParams, Redirect } from 'react-router'
import { ArrowBack } from '@material-ui/icons'
import { composeParams } from './helpers'

const useStyles = makeStyles(() =>
  createStyles({
    beerDetails: {
      display: 'flex',
      height: '100%',
      width: '100%',
      alignItems: 'center',
      justifyContent: 'center',
      flexDirection: 'column'
    },
    container: {
      display: 'flex',
      height: 'calc(100% - 56px)',
      alignItems: 'center',
      justifyContent: 'center',
      flexDirection: 'column',
      overflow: 'hidden'
    },
    content: {
      width: '100%',
      height: '100%',
      display: 'grid',
      gridTemplateRows: '2fr 1fr'
    },
    image: {
      height: 200,
      width: 200,
      backgroundSize: 'contain',
    },
    tileImage: {
      height: 100,
      width: 100,
      backgroundSize: 'contain',
    },
    similarBeers: {
      display: 'flex',
      position: 'absolute',
    },
    similarBeersWrapper: {
      position: 'relative',
      overflow: 'scroll',
      height: '100%'
    },
    similarBeer: {
      height: 200,
      width: 200,
      display: 'flex',
      justifyContent: 'center',
      alignItems: 'center'
    }
  }),
)

const Details: FC<{}> = () => {
  const { beerId } = useParams()
  const paramValid = beerId && beerId.match(/^[0-9]+$/)
  const classes = useStyles()
  const [beer, setBeer] = useState<Beer>()
  const [similarBeers, setSimilarBeers] = useState<Beer[]>()
  const [newSelectedBeer, setNewSelectedBeer] = useState<number>()
  const [requestPending, setRequestPending] = useState(true)
  const [errorMessage, setErrorMessage] = useState('')
  const [back, setBack] = useState(false)

  const handleToastClose = () => setErrorMessage('')

  useEffect(() => {
    setRequestPending(true)
    beerId && paramValid && callApi<Beer[]>(`beers/${beerId}`, RequestMethod.GET)
      .then((beers) => {
        const beerDetails = beers[0]
        setBeer(beerDetails)
        const composedParams = composeParams(
          ['ibu_gt', `${beerDetails.ibu - 1}`],
          ['ibu_lt', `${beerDetails.ibu + 1}`]
        )
        return beerDetails && callApi<Beer[]>(`beers?${composedParams}`, RequestMethod.GET)
      })
      .then((newSimilarBeers) => setSimilarBeers(newSimilarBeers.filter((beer) => beer.id !== parseInt(beerId))))
      .catch((status) => status !== 404 && setErrorMessage('Something went wrong'))
      .finally(() => setRequestPending(false))
  }, [beerId])

  const handleArrowButtonClick = () => setBack(true)
  const handleSimilarBeerClick = (beerId: number) => () => setNewSelectedBeer(beerId)

  return (
    <>
      {(!paramValid || back) && <Redirect to="/" />}
      {newSelectedBeer && <Redirect to={`/beer/${newSelectedBeer}`} />}
      <AppBar position="relative">
        <Toolbar disableGutters>
          <IconButton onClick={handleArrowButtonClick}>
            <ArrowBack />
          </IconButton>
          <Typography
            align="center"
          >
            Details
          </Typography>
        </Toolbar>
      </AppBar>
      <div className={classes.container}>
        {requestPending ? (
          <CircularProgress />
        ) : beer && similarBeers ? (
          <div className={classes.content}>
            <Card className={classes.beerDetails} key={beer.name}>
              <CardMedia
                className={classes.image}
                image={beer.image_url}
              />
              <CardContent>
                <Typography component="h5" variant="h5">
                  {beer.name}
                </Typography>
                <Typography variant="subtitle1" color="textSecondary">
                  {"Description: "}
                  {beer.description}
                </Typography>
                <Typography variant="subtitle1" color="textSecondary">
                  {"Berwer tips: "}
                  {beer.brewers_tips}
                </Typography>
                <Typography variant="subtitle1" color="textSecondary">
                  {"IBU: "}
                  {beer.ibu}
                </Typography>
                <Typography variant="subtitle1" color="textSecondary">
                  {"ABV: "}
                  {beer.abv}
                </Typography>
              </CardContent>
            </Card>
            <div>
              <Typography>Similar beers: </Typography>
              <div className={classes.similarBeersWrapper}>
                <Paper className={classes.similarBeers}>
                  {similarBeers.map((similarBeer) => (
                    <Card key={similarBeer.id} className={classes.similarBeer} onClick={handleSimilarBeerClick(similarBeer.id)}>
                      <CardMedia
                        className={classes.tileImage}
                        image={similarBeer.image_url}
                      />
                      {similarBeer.name}
                    </Card>
                  ))}
                </Paper>
              </div>
            </div>
          </div>
        ) : (
              <Typography>404 <span role="img" aria-label="beer">üç∫</span></Typography>
            )
        }
        <Toast
          open={!!errorMessage}
          onClose={handleToastClose}
          type="error"
        >
          {errorMessage}
        </Toast>
      </div>
    </>
  )
}

export default Details
