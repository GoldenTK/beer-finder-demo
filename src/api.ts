const apiUrl = 'https://api.punkapi.com/v2'

export enum RequestMethod {
  GET = 'GET',
  POST = 'POST',
  PATCH = 'PATCH'
}

const callApi = <R> (endpoint: string, requestMethod: RequestMethod, requestBody?: Object) =>
  fetch(
    `${apiUrl}/${endpoint}`, 
    { 
      method: requestMethod, 
      body: JSON.stringify(requestBody), 
      headers: {
        'Content-Type': 'application/json'
    } }
  )
    .then(res =>
      res.ok
        ? res.json() as Promise<R>
        : Promise.reject(res.status))

export default callApi
