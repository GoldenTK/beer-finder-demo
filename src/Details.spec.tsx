import { stub } from 'sinon'
import { shallow, mount } from 'enzyme'
import React from 'react'
import Details from './Details'
import reactRouter from 'react-router'
import { Card, CardContent, Typography } from '@material-ui/core'
import * as callApi from './api'
import { act } from 'react-dom/test-utils'

describe('Details', () => {
  const beer = {
    id: 0,
    name: 'fakeBeer',
    description: 'fakeDescription',
    image_url: 'fakeUrl',
    brewers_tips: '',
    ibu: 12,
    abv: 12,
  }
  const similarBeer = {
    id: 1,
    name: 'fakeBeer',
    description: 'fakeDescription',
    image_url: 'fakeUrl',
    brewers_tips: '',
    ibu: 12,
    abv: 12,
  }
  
  it('renders correct structure', () => {
    const fakeUseParams = stub(reactRouter, 'useParams').returns({ beerId: '1' })
    const details = shallow(<Details />)
    expect(details).toMatchSnapshot()
    fakeUseParams.restore()
  })
  
  it('should render beer details', async () => {
    const fakeUseParams = stub(reactRouter, 'useParams').returns({ beerId: '1' })
    const apiStub = stub(callApi, 'default')
      .onFirstCall().returns(Promise.resolve([beer]))
      .onSecondCall().returns(Promise.resolve([similarBeer]))
    const details = mount(<Details />)
    await act(async () => {
      await Promise.resolve(details);
      await new Promise(resolve => setImmediate(resolve));
      details.update();
    });
    const typographyList = details.find(Card).find(CardContent).find(Typography)
  
    expect(typographyList.length).toEqual(5)
    expect(typographyList.at(0).children().text()).toEqual(beer.name)
    expect(typographyList.at(1).children().text()).toEqual(`Description: ${beer.description}`)
    expect(typographyList.at(2).children().text()).toEqual(`Berwer tips: ${beer.brewers_tips}`)
    expect(typographyList.at(3).children().text()).toEqual(`IBU: ${beer.ibu}`)
    expect(typographyList.at(4).children().text()).toEqual(`ABV: ${beer.abv}`)
  
    fakeUseParams.restore()
    apiStub.restore()
  })
})
