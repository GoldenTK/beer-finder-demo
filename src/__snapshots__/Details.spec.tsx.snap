// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Details renders correct structure 1`] = `ShallowWrapper {}`;

exports[`renders correct structure 1`] = `ShallowWrapper {}`;
